include/sys/onoff.h

```diff
@@ -568,44 +565,119 @@ int onoff_service_reset(struct onoff_service *srv,
+ * manager will initiate a transition to off.  If there are pending
+ * requests when a transition to off completes the manager will
+ * initiate a transition to on.
  *
- * @param srv the service for which an operation is to be cancelled.
+ * @param mgr the manager for which an operation is to be cancelled.
  *
  * @param cli a pointer to the same client state that was provided
- * when the operation to be cancelled was issued.  If the cancellation
- * is successful the client will be notified of operation completion
- * with a result of `-ECANCELED`.
+ * when the operation to be cancelled was issued.
  *
  * @retval 0 if the cancellation was completed before the client could
- * be notified.  The client will be notified through cli with an
- * operation completion of `-ECANCELED`.
+ * be notified.  The cancellation succeeds and control of the cli
+ * structure returns to the client without any completion notification.
+ * @retval 1 in the case where a pending release was cancelled before
+ * the service transitioned to off.  In this case the manager retains
+ * control of the client structure, which is converted to a request,
+ * and completion will be notified when the manager completes a
+ * transition back to on.
  * @retval -EINVAL if the parameters are invalid.
- * @retval -EWOULDBLOCK if cancellation was rejected because the
- * client is the only waiter for an in-progress transition.  The
- * service retains control of the client structure.
  * @retval -EALREADY if cli was not a record of an uncompleted
  * notification at the time the cancellation was processed.  This
  * likely indicates that the operation and client notification had
  * already completed.
  */
-int onoff_cancel(struct onoff_service *srv,
+int onoff_cancel(struct onoff_manager *mgr,
 		 struct onoff_client *cli);
 
+/**
+ * @brief Signature used to notify a monitor of an onoff service of
+ * errors or completion of a state transition.
+ *
+ * This is similar to onoff_client_callback but provides information
+ * about all transitions, not just ones associated with a specific
+ * client.
+ *
+ * These functions may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-callable and non-suspendable.
+ *
+ * The callback is permitted to unregister itself from the manager,
+ * but must not register or unregister any other monitors.
+ *
+ * @param mgr the manager for which a transition has completed.
+ *
+ * @param state the state of the machine at the time of completion,
+ * restricted by ONOFF_STATE_MASK.
+ *
+ * @param res the result of the operation.  Expected values are
+ * service- and state-specific, but the value shall be non-negative if
+ * the operation succeeded, and negative if the operation failed.
+ */
+typedef void (*onoff_monitor_callback)(struct onoff_manager *mgr,
+				       struct onoff_monitor *mon,
+				       u32_t state,
```

Comment:
does it report also transient states (to off, to on)? if not maybe it could be just bool (true means on)? Then ONOFF_STATE_MASK could be private in source file.

Reply:
Yes, it reports initiation of a transition.  This can be important because at that point the service can't be used anymore; somebody might watch that and choose to not initiate a request if it's known the service is shutting down.

State mask also covers the error bit, which is public.